### PR TITLE
*: add `Reader`, `Transformer` for relay log

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

this PR is one part of #91 

### What is changed and how it works?

1. add a `MockReader` used for testing
2. add a `FileReader` used for:
	1. read relay log by syncer unit later
	2. read relay log for recovering/rollbacking an incomplete relay log file
3. add a `GenRotateEvent` to generate `RoateEvent` for testing
4. add a `Reader` (wrap `TCPReader`) used to read binlog events from master
	1. handle some errors
	2. do retry if possible
5. add a `Transformer` used to transform binlog events
	1. extract binlog position/GTID
	2. decide the event whether needed to be written to relay log

### Check List 

Tests 

 - Unit test

---

for relay log, at least following things needed to do:
1. add a `Writer` used to write binlog events to relay log file
2. compose `Reader`, `Transformer` and `Writer` together to pull binlog events from a master server and write them to files
3. wrap them to support master-slave switch and clean old code

I'll do them in some PRs later.

the flow of relay log (without master-slave switch) will be:
1. `TCPReader`read binlog events from the master
2. `Reader` handle possible errors
3. `Transfomer` get events from `Reader` and transform them
4. `Writer` get events from `Transformer` and write them to disk (with a `FileWriter`)

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@GregoryIan @amyangfei PTAL

File: pkg/binlog/event/event.go [hunk 1]
```diff
@@ -159,6 +159,33 @@ func GenFormatDescriptionEvent(header *replication.EventHeader, latestPos uint32
 	return ev, errors.Trace(err)
 }
 
+// GenRotateEvent generates a RotateEvent.
+// ref: https://dev.mysql.com/doc/internals/en/rotate-event.html
+func GenRotateEvent(header *replication.EventHeader, latestPos uint32, nextLogName []byte, position uint64) (*replication.BinlogEvent, error) {
```
Comment (GregoryIan) [inline pkg/binlog/event/event.go]: what's difference between `latestPos` and `position`?

Comment (csuzhangxc) [reply pkg/binlog/event/event.go]: `latestPos` is the previous end position of the latest events or the start position of _this_ event.
`position` is the rotate target position for _this_ event or the start position of the _next_ event.

File: relay/transformer/transformer.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+var (
+	artificialFlag = uint16(0x0020) // LOG_EVENT_ARTIFICIAL_F
+)
+
+// Result represents a transform result.
+type Result struct {
+	Ignore      bool          // whether the event should be ignored
+	LogPos      uint32        // binlog event's End_log_pos or Position in RotateEvent
+	NextLogName string        // next binlog filename, only valid for RotateEvent
+	GTIDSet     mysql.GTIDSet // GTIDSet got from QueryEvent and XIDEvent when RawModeEnabled not true
```
Comment (GregoryIan) [inline relay/transformer/transformer.go]: it's gtid or gtidset?

File: relay/transformer/transformer.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	CanSaveGTID bool          // whether can save GTID into meta, true for DDL query and XIDEvent
+}
+
+// Transformer receives binlog events from a reader and transforms them.
+// The transformed binlog events should be send to one or more writers.
+// The transformer should support:
+//   1. extract binlog position, GTID info from the event.
+//   2. decide the event whether needed by a downstream writer.
+//     - the downstream writer may also drop some events according to its strategy.
+// NOTE: more works maybe moved from outer into Transformer later.
```
Comment (GregoryIan) [inline relay/transformer/transformer.go]:
```suggestion
// NOTE: more features maybe moved from outer into Transformer later.
```

File: relay/transformer/transformer.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+)
+
+var (
+	artificialFlag = uint16(0x0020) // LOG_EVENT_ARTIFICIAL_F
+)
+
+// Result represents a transform result.
+type Result struct {
```
Comment (GregoryIan) [inline relay/transformer/transformer.go]: I don't understand this design very well,  like
* if some binlog is ignored, should writer write it?
* what's its transform rules?

File: pkg/binlog/reader/mock_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+			obtained = append(obtained, testMockCase{ev: nil, err: err})
+		} else {
+			obtained = append(obtained, testMockCase{ev: ev, err: nil})
+		}
+		if len(obtained) == len(cases) || ctx.Err() != nil {
+			break
+		}
+	}
+
+	c.Assert(ctx.Err(), IsNil)
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/binlog/reader/mock_test.go]: `|| ctx.Err() != nil` is used to break the `for` loop if there are bugs.
if timemout, then the test should be failed in this `c.Assert(ctx.Err(), IsNil)`.

File: pkg/binlog/reader/file.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	r.ctx, r.cancel = context.WithCancel(context.Background())
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		err := r.parser.ParseFile(pos.Name, int64(pos.Pos), r.onEvent)
+		if err != nil {
+			select {
+			case r.ech <- err:
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/binlog/reader/file.go]: addressed. log the error before the `select`.

File: pkg/binlog/reader/file.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+	select {
+	case ev := <-r.ch:
+		r.sendOffset.Set(ev.Header.LogPos)
+		return ev, nil
+	case err := <-r.ech:
+		return nil, err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-r.ctx.Done(): // Reader closed
+		return nil, r.ctx.Err()
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/binlog/reader/file.go]: added.

File: relay/reader/reader.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+func (r *reader) GetEvent(ctx context.Context) (*replication.BinlogEvent, error) {
+	if r.stage.Get() != int32(stagePrepared) {
+		return nil, errors.Errorf("stage %s, expect %s", readerStage(r.stage.Get()), stagePrepared)
+	}
+
+	for {
+		ev, err := r.in.GetEvent(ctx)
+		if err == nil {
+			return ev, nil
+		} else if isRetryableError(err) {
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply relay/reader/reader.go]: removed timeout error handing in this reader, that should be handled in the outer caller.

File: relay/reader/reader.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func (r *reader) setUpReaderByGTID() error {
+	gs := r.cfg.GTIDs
+	log.Infof("[relay] start sync for master %s from GTID set %s", r.cfg.MasterID, gs)
+	err := r.in.StartSyncByGTID(gs)
+	if err != nil {
+		log.Errorf("[relay] start sync for master %s from GTID set %s error %v",
+			r.cfg.MasterID, gs, errors.ErrorStack(err))
+		return r.setUpReaderByPos()
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply relay/reader/reader.go]: OK, let's put this error to be handled by the outer or event the user.

File: relay/reader/reader.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	} else {
+		err = r.setUpReaderByPos()
+	}
+
+	r.stage.Set(int32(stagePrepared))
+	return err
+}
+
+// Close implements Reader.Close.
+func (r *reader) Close() error {
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply relay/reader/reader.go]: ❤️ added. and change `Mutex` to `RWMutex`, and remove the atomic variable check

File: relay/reader/reader.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	err := r.in.Close()
+	r.stage.Set(int32(stageClosed))
+	return errors.Trace(err)
+}
+
+// GetEvent implements Reader.GetEvent.
+// If some ignorable error occurred, the returned event and error both are nil.
+func (r *reader) GetEvent(ctx context.Context) (*replication.BinlogEvent, error) {
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply relay/reader/reader.go]: added

File: relay/transformer/transformer.go (hunk 2 above)
Comment (csuzhangxc) [reply relay/transformer/transformer.go]: 👍 It's gtid set. It's GTID in one binlog event, but _go-mysql_ returns a GTID set (including the previous/passed in GTID set)

File: relay/transformer/transformer.go (hunk 4 above)
Comment (csuzhangxc) [reply relay/transformer/transformer.go]: binlog event may be ignored in multiple steps in the processing chain, at least:
1. some fake/artificial events, which will be marked as ignored in this transformer
2. some obsolete events (like FormatDescriptionEvent), which will be ignored by the writer

maybe the name of `transformer` is not accurate, and in fact, more features may be added into this later.

**Comment by csuzhangxc** [comment]
@GregoryIan PTAL again

File: pkg/binlog/reader/mock_test.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+	c.Assert(r.StartSyncByGTID(nil), IsNil)
+	c.Assert(r.Status(), IsNil)
+	c.Assert(r.Close(), IsNil)
+
+	// replace with special error
+	mockR := r.(*MockReader)
+	errSpecial := errors.New("special error for methods")
+	mockR.ErrStartByPos = errSpecial
+	mockR.ErrStartByGTID = errSpecial
+	mockR.ErrClose = errSpecial
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/binlog/reader/mock_test.go]: changed to use different errors

File: relay/reader/reader.go [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+	for {
+		ev, err := r.in.GetEvent(ctx)
+		if err == nil {
+			return ev, nil
+		} else if isIgnorableError(err) {
+			log.Warnf("[relay] get event with ignorable error %s", err)
+			return nil, nil // return without error and also without binlog event
+		}
+		// NOTE: add retryable error support if needed later
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply relay/reader/reader.go]: I moved it to L128.
The `for` loop executed exactly once now but may need to support `retryable error` later.

File: relay/reader/reader_test.go [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+	err = r.Close()
+	c.Assert(err, NotNil) // call multi times
+
+	// getEvent from a closed reader
+	ev, err := r.GetEvent(ctx)
+	c.Assert(err, NotNil)
+	c.Assert(ev, IsNil)
+}
+
+func (t *testReaderSuite) testBackOffWithReader(c *C, r Reader, errByPos error, errByGTID error) error {
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply relay/reader/reader_test.go]: forgot to remove ❤️

**Comment by csuzhangxc** [comment]
@GregoryIan @amyangfei PTAL again

File: pkg/binlog/reader/file_test.go [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+
+	// check status, stageNew
+	status := r.Status()
+	frStatus, ok := status.(*FileReaderStatus)
+	c.Assert(ok, IsTrue)
+	c.Assert(frStatus.Stage, Equals, stageNew.String())
+	c.Assert(frStatus.ReadOffset, Equals, uint32(0))
+	c.Assert(frStatus.SendOffset, Equals, uint32(0))
+	frStatusStr := frStatus.String()
+	c.Assert(strings.Contains(frStatusStr, stageNew.String()), IsTrue)
```
Comment (amyangfei) [inline pkg/binlog/reader/file_test.go]: L59-L60 can be replaced with
```
c.Assert(frStatus, Matches, fmt.Sprintf(`.*"stage":"%s".*`, stageNew))
```
is this better

File: pkg/binlog/reader/file.go [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+// Close implements Reader.Close.
+func (r *FileReader) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stage == stageClosed {
+		return errors.New("already closed")
+	}
+
+	r.parser.Stop()
```
Comment (GregoryIan) [inline pkg/binlog/reader/file.go]: it's better to put it after L135, even I know it's ok after I review its implemention

Comment (csuzhangxc) [reply pkg/binlog/reader/file.go]: addressed

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**